### PR TITLE
Add a true JSON AST serializer

### PR DESCRIPTION
**Pull request by matozoid**
We've had a bunch of issues with the JsonPrinter and its friends, which were never meant to do serialization. These new classes are.

### CODE REVIEW

File: javaparser-core-serialization/src/main/java/com/github/javaparser/serialization/JavaParserJsonDeserializer.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
+package com.github.javaparser.serialization;
```
Comment (ftomassetti) [inline javaparser-core-serialization/src/main/java/com/github/javaparser/serialization/JavaParserJsonDeserializer.java:21]: Do we need copyright statements on new files?

File: javaparser-core-serialization/src/main/java/com/github/javaparser/serialization/JavaParserJsonDeserializer.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    }
+
+    private Node deserializeObject(JsonObject nodeJson) {
+        try {
+            String serializedNodeType = nodeJson.getString("!");
+            BaseNodeMetaModel nodeMetaModel = getNodeMetaModel(Class.forName(serializedNodeType))
+                    .orElseThrow(() -> new IllegalStateException("Trying to deserialize an unknown node type: " + serializedNodeType));
+            Map<String, Object> parameters = new HashMap<>();
+            for (String name : nodeJson.keySet()) {
+                if (name.equals("!")) {
```
Comment (ftomassetti) [inline javaparser-core-serialization/src/main/java/com/github/javaparser/serialization/JavaParserJsonDeserializer.java]: What this special name mean? Do we need a constant for that?

File: javaparser-core-serialization/src/main/java/com/github/javaparser/serialization/JavaParserJsonSerializer.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    private void serialize(String nodeName, Node node, JsonGenerator generator) {
+        requireNonNull(node);
+        BaseNodeMetaModel nodeMetaModel = JavaParserMetaModel.getNodeMetaModel(node.getClass()).orElseThrow(() -> new IllegalStateException("Unknown Node: " + node.getClass()));
+
+        if (nodeName == null) {
+            generator.writeStartObject();
+        } else {
+            generator.writeStartObject(nodeName);
+        }
+        generator.write("!", node.getClass().getName());
```
Comment (ftomassetti) [inline javaparser-core-serialization/src/main/java/com/github/javaparser/serialization/JavaParserJsonSerializer.java]: Perhaps we could use a constant for that or maybe give a way to set this name in some sort of configuration

**Review by ftomassetti — APPROVED** [review]
It seems ok to me.
I think that some way of processing nodes during serialization could become handy later. But let's users ask for that :D

**Comment by matozoid** [comment]
Fixes #1379

**Pull request merged**